jerry-core/jerry-snapshot.c

```diff
@@ -630,3 +632,409 @@ jerry_exec_snapshot (const void *snapshot_p, /**< snapshot */
 /**
  * @}
  */
+
+#ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+
+/**
+ * ====================== Functions for literal saving ==========================
+ */
+
+/**
+ * Helper function for a merge sort algorithm.
+ */
+static void
+jerry_save_literals_merging_helper (ecma_string_t *src[], /**< input array */
+                                    ecma_string_t *dest[], /**< output array */
+                                    lit_utf8_size_t low, /**< begin index */
+                                    lit_utf8_size_t mid, /**< middle index */
+                                    lit_utf8_size_t high) /**< end index */
+{
+  lit_utf8_size_t x, y, i;
+
+  for (x = low, y = mid + 1, i = low; x <= mid && y <= high; i++)
+  {
+    const lit_utf8_size_t x_size = ecma_string_get_size (src[x]);
+    const lit_utf8_size_t y_size = ecma_string_get_size (src[y]);
+
+    if (x_size == y_size)
+    {
+      if (ecma_compare_ecma_strings_relational (src[x], src[y]))
+      {
+        dest[i] = src[x++];
+      }
+      else
+      {
+        dest[i] = src[y++];
+      }
+    }
+    else if (x_size < y_size)
+    {
+      dest[i] = src[x++];
+    }
+    else
+    {
+      dest[i] = src[y++];
+    }
+  }
+
+  while (x <= mid)
+  {
+    dest[i++] = src[x++];
+  }
+
+  while (y <= high)
+  {
+    dest[i++] = src[y++];
+  }
+
+  for (i = low; i <= high; i++)
+  {
+    src[i] = dest[i];
+  }
+} /* jerry_save_literals_merging_helper */
+
+/**
+ * Helper function for a merge sort algorithm.
+ */
+static void
+jerry_save_literals_sort_helper (ecma_string_t *src[], /**< input array */
+                                 ecma_string_t *dest[], /**< output array */
+                                 lit_utf8_size_t begin, /**< begin index */
+                                 lit_utf8_size_t end) /**< end index */
+{
+  lit_utf8_size_t mid = (begin + end) / 2;
+
+  if (begin < end)
+  {
+    jerry_save_literals_sort_helper (src, dest, begin, mid);
+    jerry_save_literals_sort_helper (src, dest, mid + 1, end);
+    jerry_save_literals_merging_helper (src, dest, begin, mid, end);
```

Comment:
Not a complicated algorithm for an array of data:

https://rosettacode.org/wiki/Sorting_algorithms/Heapsort#C